service/topology_coordinator.cc

```diff
@@ -664,6 +665,119 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
         }
     }
 
+    template<typename HostIdToDC>
+    static void downsize_tablet_replicas_to_rf(locator::tablet_replica_set& replicas, size_t current_rf,
+        size_t new_rf, const locator::load_sketch& load, const sstring& dc, HostIdToDC host_id_to_dc)
+    {
+        using host_idx_to_load_t = std::pair<size_t, double>;
+        std::vector<host_idx_to_load_t> host_load;
+        host_load.reserve(replicas.size());
+
+        for (size_t i = 0; i < replicas.size(); i++) {
+            host_idx_to_load_t host_load_pair;
+            if (host_id_to_dc(replicas[i].host) == dc) {
+                host_load_pair.first = i;
+                host_load_pair.second = load.get_load(replicas[i].host);
+                host_load.push_back(host_load_pair);
+            }
+        }
+
+        auto cmp_by_load = [](const auto& a, const auto& b) {
+            return a.second > b.second;
+        };
+        auto cmp_by_idx = [](const auto& a, const auto& b) {
+            return a.first > b.first;
+        };
+
+        size_t to_remove = current_rf - new_rf;
+
+        // ensure that the most loaded replicas are in the beginning
+        std::nth_element(host_load.begin(), host_load.begin() + to_remove,
+            host_load.end(), cmp_by_load);
+
+        // sort the most loaded replicas by id, descending so that
+        // we can remove them by their index
+        std::sort(host_load.begin(), host_load.begin() + to_remove, cmp_by_idx);
+
+        for (size_t i = 0; i < to_remove; i++) {
+            replicas.erase(replicas.begin() + host_load[i].first);
+        }
+    }
+
+    static void upsize_tablet_replicas_to_rf(locator::tablet_replica_set& replicas, size_t current_rf,
+        size_t new_rf, locator::load_sketch& load, const locator::token_metadata_ptr& tm, const sstring& dc)
+    {
+        for (const auto& endpoint : tm->get_topology().get_datacenter_endpoints().at(dc)) {
+            locator::host_id ep = tm->get_topology().find_node(endpoint)->host_id();
+            replicas.emplace_back(locator::tablet_replica{ep, load.next_shard(ep)});
```

Comment:
This can violate RF constraints:
  - no two replicas on the same node (this can allocate a node which is already a replica)
  - if possible, spread replicas to different racks (to maintain high availability)

Reply:
these are good candidates for assertions